### PR TITLE
nrf_security: Add sanity check for psa key attributes

### PR DESCRIPTION
**Pull request by mswarowsky**
When calling psa_generate_key/import_key/copy_key an key attributes description is passed, wich contains values like, key_type, key_algorithm, etc. This adds a sanity check to ensure key type, algorithm and size are a valid combination according to the psa crypto spec.

Ref: NCSDK-17103

Signed-off-by: Markus Swarowsky <markus.swarowsky@nordicsemi.no>

### CODE REVIEW

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 1]
```diff
@@ -97,6 +97,216 @@
...
+
+/**
+ * @brief This can be used to extract just the base value of algorithms that
+ *  are combined with a hashing scheme. 
+ * @param alg The full algorithm identifier.
+ */
+#define GET_ALG_BASE(alg) (alg & (~PSA_ALG_HASH_MASK))
+
+/**
+ * @brief Compare AEAD alogrithms, by ingoring the the output tag length
```
Comment (vili-nordic) [inline nrf_security/src/psa_crypto_driver_wrappers.c:109]: algorithms. ignoring

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 2]
```diff
@@ -97,6 +97,216 @@
...
+ * 
+ */
+#define EQUAL_AEAD_ALG(alg_1, alg_2) ((alg_1 & ~PSA_ALG_AEAD_TAG_LENGTH_MASK) \
+                                == (alg_2 & ~PSA_ALG_AEAD_TAG_LENGTH_MASK)) 
+
+/**
+ * @brief Checks if the given algorithm is one the follwing ECDSA algorithms
```
Comment (vili-nordic) [inline nrf_security/src/psa_crypto_driver_wrappers.c:116]: following

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 3]
```diff
@@ -97,6 +97,216 @@
...
+ */
+static bool is_ecdsa_or_ecdh(psa_algorithm_t algorithm){
+    return (GET_ALG_BASE(algorithm) == PSA_ALG_DETERMINISTIC_ECDSA_BASE  ||
+           GET_ALG_BASE(algorithm) == PSA_ALG_ECDSA_BASE ||
+           algorithm == PSA_ALG_ECDH); // (key pair only)
+}
+
+/**
+ * @brief Checks if the given combination of key curve, bit size and algorithm 
+ *  is a valid tuple by the PSA standart
```
Comment (vili-nordic) [inline nrf_security/src/psa_crypto_driver_wrappers.c:130]: standart

**Review by vili-nordic — COMMENTED** [review]
looks ok to me. Some spelling errors.

Commit message:
*an key
*wich

**Review by vili-nordic — APPROVED** [review]

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 4]
```diff
@@ -97,6 +98,216 @@
...
+
+/**
+ * @brief Checks if the given algorithm is one the following ECDSA algorithms
+ * PSA_ALG_DETERMINISTIC_ECDSA
+ * PSA_ALG_ECDSA
+ * PSA_ALG_ECDSA_ANY PSA_ALG_ECDH
+ * or the PSA_ALG_ECDH 
+ */
+static bool is_ecdsa_or_ecdh(psa_algorithm_t algorithm){
+    return (GET_ALG_BASE(algorithm) == PSA_ALG_DETERMINISTIC_ECDSA_BASE  ||
```
Comment (Vge0rge) [inline nrf_security/src/psa_crypto_driver_wrappers.c:124]: I think that it is better to use the defined macros from PSA.
Look at: PSA_ALG_ECDSA_IS_DETERMINISTIC(alg)
PSA_ALG_IS_ECDSA(alg)
PSA_ALG_IS_ECDH(alg)

They are basically doing the same thing.

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 5]
```diff
@@ -97,6 +98,216 @@
...
+    }
+
+    return PSA_ERROR_INVALID_ARGUMENT;
+}
+
+/**
+ * @brief Checks if the given combination of key type, bit size and algorithm 
+ *  is a valid tuple by the PSA crypto spec. 
+ *  
+ * @return psa_status_t psa_status_t PSA_ERROR_INVALID_ARGUMENT if invalid else PSA_SUCCESS
```
Comment (Vge0rge) [inline nrf_security/src/psa_crypto_driver_wrappers.c:214]: Remove one psa_status_t

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 6]
```diff
@@ -97,6 +98,216 @@
...
+
+    if( bits % 8 != 0 ) {
+        return( PSA_ERROR_INVALID_ARGUMENT );
+    }
+
+    switch( type )
+    {
+        case PSA_KEY_TYPE_RAW_DATA:
+        case PSA_KEY_TYPE_DERIVE:
+            if(GET_ALG_BASE(algorithm) == PSA_ALG_HKDF_BASE ||
```
Comment (Vge0rge) [inline nrf_security/src/psa_crypto_driver_wrappers.c:233]: There is a macro for and for the TLS12 functions.
You can grep "PSA_ALG_IS" in the include/psa/ and you will see all the available macros.

**Review by Vge0rge — CHANGES_REQUESTED** [review]

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 7]
```diff
@@ -97,6 +98,216 @@
...
+
+/**
+ * @brief This can be used to extract just the base value of algorithms that
+ *  are combined with a hashing scheme. 
+ * @param alg The full algorithm identifier.
+ */
+#define GET_ALG_BASE(alg) (alg & (~PSA_ALG_HASH_MASK))
+
+/**
+ * @brief Compare AEAD alogrithms ingoring the the output tag length
```
[the remark this answers is not in the document]
Comment (vili-nordic) [reply nrf_security/src/psa_crypto_driver_wrappers.c:110]: and alogrithms
and ingoring

File: nrf_security/src/psa_crypto_driver_wrappers.c [hunk 8]
```diff
@@ -97,6 +98,208 @@
 #include "psa_crypto_se.h"
 #endif
 
+/**
+ * @brief Compare AEAD alogrithms and ingoring the output tag length
```
Comment (vili-nordic) [inline nrf_security/src/psa_crypto_driver_wrappers.c:102]: * algorithms
* ignoring

**Review by frkv — APPROVED** [review]
LGTM

**Pull request closed without merging**